src/rest_vol.c

```diff
@@ -2307,7 +2360,209 @@ RV_find_object_by_path(RV_object_t *parent_obj, const char *obj_path, H5I_type_t
     } /* end if */
 
     return ret_value;
-} /* end RV_find_object_by_path() */
+} /* end RV_find_object_by_path1() */
+
+/*-------------------------------------------------------------------------
+ * Function:    RV_find_object_by_path2
+ *
+ * Purpose:     Given a pathname, this function is responsible for making
+ *              an HTTP GET request to the server in order to retrieve
+ *              information about an object.
+ *
+ *              This function uses the h5path parameter in
+ *              its request, so if the provided path is to a symbolic link,
+ *              the server will follow that link until it encounters the
+ *              final object by hard link, and respond with information about
+ *              that object.
+ *
+ *              Previously, the approach to finding an object was to
+ *              recursively traverse links in the file, but this caused
+ *              too much communication between client and server and would
+ *              start to become problematic for deeply-nested objects.
+ *
+ *              The type of the retrieved object will be stored at the
+ *              provided target_object_type field.
+ *
+ * Return:      Non-negative on success, negative on failure
+ *
+ * Programmer:  Jordan Henderson
+ *              November, 2017
+ */
+htri_t
+RV_find_object_by_path2(RV_object_t *parent_obj, const char *obj_path, H5I_type_t *target_object_type,
+                        herr_t (*obj_found_callback)(char *, void *, void *), void *callback_data_in,
+                        void *callback_data_out)
```

Comment:
There's a lot of duplication between the two versions of this function that probably could have been handled by just checking the server version and determining the correct URL to use and whether or not to make use of the RV_parse_type callback, but we can worry about cleaning this up later.